File: src/frontend/cte/lib/app/data/model/cte_ferroviario_vagao_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cte/app/data/model/model_imports.dart';
import 'package:cte/app/data/domain/domain_imports.dart';

class CteFerroviarioVagaoModel {
	int? id;
	int? idCteFerroviario;
	int? numeroVagao;
	double? capacidade;
	String? tipoVagao;
	double? pesoReal;
	double? pesoBc;
	CteFerroviarioModel? cteFerroviarioModel;

	CteFerroviarioVagaoModel({
		this.id,
		this.idCteFerroviario,
		this.numeroVagao,
		this.capacidade,
		this.tipoVagao,
		this.pesoReal,
		this.pesoBc,
		this.cteFerroviarioModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero_vagao',
		'capacidade',
		'tipo_vagao',
		'peso_real',
		'peso_bc',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero Vagao',
		'Capacidade',
		'Tipo Vagao',
		'Peso Real',
		'Peso Bc',
	];

	CteFerroviarioVagaoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCteFerroviario = jsonData['idCteFerroviario'];
		numeroVagao = jsonData['numeroVagao'];
		capacidade = jsonData['capacidade']?.toDouble();
		tipoVagao = CteFerroviarioVagaoDomain.getTipoVagao(jsonData['tipoVagao']);
		pesoReal = jsonData['pesoReal']?.toDouble();
		pesoBc = jsonData['pesoBc']?.toDouble();
		cteFerroviarioModel = jsonData['cteFerroviarioModel'] == null ? CteFerroviarioModel() : CteFerroviarioModel.fromJson(jsonData['cteFerroviarioModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCteFerroviario'] = idCteFerroviario != 0 ? idCteFerroviario : null;
		jsonData['numeroVagao'] = numeroVagao;
		jsonData['capacidade'] = capacidade;
		jsonData['tipoVagao'] = CteFerroviarioVagaoDomain.setTipoVagao(tipoVagao);
		jsonData['pesoReal'] = pesoReal;
		jsonData['pesoBc'] = pesoBc;
		jsonData['cteFerroviarioModel'] = cteFerroviarioModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCteFerroviario = plutoRow.cells['idCteFerroviario']?.value;
		numeroVagao = plutoRow.cells['numeroVagao']?.value;
		capacidade = plutoRow.cells['capacidade']?.value?.toDouble();
		tipoVagao = plutoRow.cells['tipoVagao']?.value != '' ? plutoRow.cells['tipoVagao']?.value : 'AAA';
		pesoReal = plutoRow.cells['pesoReal']?.value?.toDouble();
		pesoBc = plutoRow.cells['pesoBc']?.value?.toDouble();
		cteFerroviarioModel = CteFerroviarioModel();
		cteFerroviarioModel?.fluxo = plutoRow.cells['cteFerroviarioModel']?.value;
	}	

	CteFerroviarioVagaoModel clone() {
		return CteFerroviarioVagaoModel(
			id: id,
			idCteFerroviario: idCteFerroviario,
			numeroVagao: numeroVagao,
			capacidade: capacidade,
			tipoVagao: tipoVagao,
			pesoReal: pesoReal,
			pesoBc: pesoBc,
		);			
	}

	
}

File: backend/cte/src/model/cte_ferroviario_vagao_model.py
from src import db
from src.model.cte_ferroviario_model import CteFerroviarioModel


class CteFerroviarioVagaoModel(db.Model):
    __tablename__ = 'cte_ferroviario_vagao'

    id = db.Column(db.Integer, primary_key=True)
    numero_vagao = db.Column(db.Integer)
    capacidade = db.Column(db.Float)
    tipo_vagao = db.Column(db.String(3))
    peso_real = db.Column(db.Float)
    peso_bc = db.Column(db.Float)
    id_cte_ferroviario = db.Column(db.Integer, db.ForeignKey('cte_ferroviario.id'))

    cte_ferroviario_model = db.relationship('CteFerroviarioModel', foreign_keys=[id_cte_ferroviario])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_cte_ferroviario = data.get('idCteFerroviario')
        self.numero_vagao = data.get('numeroVagao')
        self.capacidade = data.get('capacidade')
        self.tipo_vagao = data.get('tipoVagao')
        self.peso_real = data.get('pesoReal')
        self.peso_bc = data.get('pesoBc')

    def serialize(self):
        return {
            'id': self.id,
            'idCteFerroviario': self.id_cte_ferroviario,
            'numeroVagao': self.numero_vagao,
            'capacidade': self.capacidade,
            'tipoVagao': self.tipo_vagao,
            'pesoReal': self.peso_real,
            'pesoBc': self.peso_bc,
            'cteFerroviarioModel': self.cte_ferroviario_model.serialize() if self.cte_ferroviario_model else None,
        }